nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,459 @@
+		return nil, err
+	}
+	return newClients, nil
+}
+
+func configureNewClient(newClient *nsxtClients, oldClient *nsxtClients, host string, username string, password string) error {
+	newClient.Host = host
+	securityCtx, err := getConfiguredSecurityContext(newClient, "", "", "", username, password)
+	if err != nil {
+		return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+	}
+	newClient.PolicySecurityContext = securityCtx
+	newClient.PolicyHTTPClient = oldClient.PolicyHTTPClient
+	err = initNSXVersion(getPolicyConnector(*newClient))
+	if err != nil {
+		return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+	}
+	return nil
+}
+
+func joinNodeToCluster(clusterID string, certSha256Thumbprint string, guestNode NsxClusterNode, masterNodeIP string, d *schema.ResourceData, m interface{}) error {
+	c, err := getNewNsxtClient(guestNode, d, m)
+	if err != nil {
+		return err
+	}
+	newNsxClients := c.(nsxtClients)
+	connector := getPolicyConnector(newNsxClients)
+	client := nsx.NewClusterClient(connector)
+	username, password := getHostCredential(m)
+	joinClusterParams := nsxModel.JoinClusterParameters{
+		CertificateSha256Thumbprint: &certSha256Thumbprint,
+		ClusterId:                   &clusterID,
+		IpAddress:                   &masterNodeIP,
+		Username:                    &username,
+		Password:                    &password,
+	}
+	_, err = client.Joincluster(joinClusterParams)
+	if err != nil {
+		return fmt.Errorf("Failed to join node to cluster: %s, node ip address: %s", err, guestNode.IPAddress)
+	}
+	return nil
+}
+
+func getHostCredential(m interface{}) (string, string) {
+	nsxtClient := m.(nsxtClients)
+	username := nsxtClient.NsxtClientConfig.UserName
+	password := nsxtClient.NsxtClientConfig.Password
+	return username, password
+}
+
+func resourceNsxtManagerClusterRead(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	clusterConfig, err := client.Get()
+	if err != nil {
+		return fmt.Errorf("Failed to read Nsxt Manager Cluster: %s", err)
+	}
+	nodeInfo := clusterConfig.Nodes
+	var nodes []NsxClusterNode
+	hostIP, err := getHostIPFromClusterConfig(m, clusterConfig)
+	isIPv4 := (net.ParseIP(hostIP)).To4() != nil
+	if err != nil {
+		return err
+	}
+	for _, node := range nodeInfo {
+		ip := getIPFromNodeInfo(node, isIPv4)
+		fqdn := *node.Fqdn
+		status := *node.Status
+		if ip != hostIP {
+			id := *node.NodeUuid
+			clusterNode := NsxClusterNode{
+				ID:        id,
+				IPAddress: ip,
+				Fqdn:      fqdn,
+				Status:    status,
+			}
+			nodes = append(nodes, clusterNode)
+		}
+
+	}
+	d.Set("revision", clusterConfig.Revision)
+	setNodesInSchema(d, nodes)
+	return nil
+}
+
+func getIPFromNodeInfo(node nsxModel.ClusterNodeInfo, isIPv4 bool) string {
+	// After join node into cluster, apiListen address may take some time to become available
+	// if it's not available, we retrieve node ip from nodeEntityInfo
```

Comment:
Are we expecting this problem to be solved with retry? If yes, can this logic be removed?

Reply:
So I think it can be solved, but here we don't need to retry because node ip is available in nodeEntityInfo, it could save some time.
Unlike getting API CertSha256Thumbprint, we have to retry to wait for APIListenAddress because it's the only way to obtain the correct thumbprint

Reply:
I would suggest to always grab the IP from https entity then, because its would mean less code and more deterministic behavior

Reply:
makes sense, I've made the changes